cmd/debug/segment-downloader/main.go

```diff
@@ -0,0 +1,241 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/vivint/infectious"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/macaroon"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls/tlsopts"
+	"storj.io/storj/pkg/rpc"
+	"storj.io/storj/uplink/ecclient"
+	"storj.io/storj/uplink/eestream"
+	"storj.io/storj/uplink/metainfo"
+	"storj.io/storj/uplink/storage/segments"
+)
+
+var (
+	flagAPIKey        = flag.String("api-key", "", "api key")
+	flagSatelliteAddr = flag.String("satellite-addr", "", "satellite address")
+)
+
+func main() {
+	flag.Parse()
+	err := run(context.Background())
+	if err != nil {
+		panic(err)
+	}
+}
+
+func run(ctx context.Context) (err error) {
+	apiKey, err := macaroon.ParseAPIKey(*flagAPIKey)
+	if err != nil {
+		return err
+	}
+
+	ident, err := identity.NewFullIdentity(ctx, identity.NewCAOptions{
+		Difficulty:  9,
+		Concurrency: 1,
+	})
+	if err != nil {
+		return err
+	}
+	tlsOpts, err := tlsopts.NewOptions(ident, tlsopts.Config{
+		PeerIDVersions: "0",
+	}, nil)
+	if err != nil {
+		return err
+	}
+
+	dialer := rpc.NewDefaultDialer(tlsOpts)
+
+	conn, err := dialer.DialAddressInsecureBestEffort(ctx, *flagSatelliteAddr)
+	if err != nil {
+		return err
+	}
+	defer func() { err = errs.Combine(err, conn.Close()) }()
+	client := conn.MetainfoClient()
+
+	ec := ecclient.NewClient(zap.L(), dialer, int(32*memory.MB))
+
+	return newDownloader(client, ec, apiKey).Audit(ctx)
+}
+
+type downloader struct {
+	client pb.MetainfoClient
+	ec     ecclient.Client
+	apiKey *macaroon.APIKey
+}
+
+func newDownloader(client pb.MetainfoClient, ec ecclient.Client, apiKey *macaroon.APIKey) *downloader {
+	return &downloader{client: client, ec: ec, apiKey: apiKey}
+}
+
+func (d *downloader) Audit(ctx context.Context) error {
+	segments, err := d.list(ctx)
+	if err != nil {
+		return err
+	}
+
+	return d.sanityCheck(ctx, segments)
+}
+
+func (d *downloader) list(ctx context.Context) (map[string]map[string]bool, error) {
+	// assume we can hold all of a project's segment paths in memory.
+	// safe assumption for now but unlikely to be true after launch
+	segments := map[string]map[string]bool{}
+
+	cursor := ""
+	for {
+		resp, err := d.client.AdminSegmentAudit(ctx, &pb.ListSegmentsRequestOld{
+			Header: &pb.RequestHeader{
+				ApiKey: d.apiKey.SerializeRaw(),
+			},
+			StartAfter: []byte(cursor),
+			Recursive:  true,
+			Limit:      100,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, item := range resp.GetItems() {
+			cursor := string(item.GetPath())
+			parts := strings.SplitN(cursor, "/", 2)
```

Comment:
I'm surprised we don't have a library function that can take a path and return its bucketpath and segment for us. This feels like a leaking implementation detail.

Then again ... I suppose path format is a pretty stable part of our api. So probably ignore this comment. :)

Reply:
The reason it's not exposed is that we shouldn't be using it. For multipart uploads/downloads we will probably need non-contiguous indices. Hence, it makes more sense for most of the codebase to rely on listing the segments and then using them as the reference.